### PR TITLE
#1647 Dependency update

### PR DESCRIPTION
**Pull request by mchlwellman**
# Description

Dependencies update - 

- Generated requirements.txt from requirements-app.txt. Notable upgrades:
  - greenlet upgraded from 2.0.2 to 3.0.3
  - gunicorn upgraded from git commit to 21.2.0
  - Flask upgraded from 2.3.3 to 3.0.2
  - pypdf upgraded from 3.17.2 to 4.0.2
  - redis upgraded from 4.6 to 5.0.1
  - Werkzeug upgraded from 2.3.8 to 3.0.1
- updated github actions
  - amazon-ecr-login
  - configure-aws-credentials
  - cache
  - setup-python
 
The following dependencies were NOT updated:

- urllib3 (pinned by botocore)
- phonenumbers (pinned by notification-utils)
- marshmallow (explicitly called out in ticket to not update)
- ddtrace (handled in separate ticket) 
- dnspython (caused errors)

Issue #1647

## Type of change

Please check the relevant option(s).

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

How has this been tested? (e.g. Unit tests, Tested locally, Tested as a GitHub action, Depoyed to dev, etc)  
Please provide relevant information and / or links to demonstrate the functionality or fix. (e.g. screenshots, link to deployment, regression test results, etc)

- Local tests pass
- PR tests pass
- Regression tests in dev pass

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket is now moved into the DEV test column

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -34,7 +34,7 @@ check-dependencies: install-safety ## Scan dependencies for security vulnerabili
 	# 12 Dec 2023: 51668 is fixed with >= 2.0.0b1 of SQLAlchemy. Ongoing refactor to upgrade.
 	# 22 Feb 2024: 64227 is fixed with >= 3.1.3 of jinja2. Update is being worked in #1647
```
Comment (cris-oddball) [inline Makefile:35]: This comment is no longer necessary now that you've removed 64227, please remove the comment for Jinja

**Review by cris-oddball — CHANGES_REQUESTED** [review]
I don't see that requirements-app was updated to reflect all the changes in requirements.txt

please go through requirements-app thoroughly and make sure that the major.monir.patch matches what you have in requirements for each dependency in requirements-app.

File: Makefile (hunk 1 above)
Comment (mchlwellman) [reply Makefile]: Ok, removed that comment

**Review by MackHalliday — APPROVED** [review]
Audited thru Dependency Bot PRs and checked double updates in PR. 

All updates we were able to make were made in this PR. 

Screenshot 2024-02-23 at 12 31 20 PM

**Comment by mchlwellman** [comment]
> I don't see that requirements-app was updated to reflect all the changes in requirements.txt
> 
> please go through requirements-app thoroughly and make sure that the major.monir.patch matches what you have in requirements for each dependency in requirements-app.

I've updated requirements-app.txt with the latest versions.

**Review by cris-oddball — APPROVED** [review]
LGTM

File: README.md [hunk 2]
```diff
@@ -259,14 +259,13 @@ To update the images, change the `FROM` directive at the top of Dockerfiles and
...
 
-1. If necessary, build the notification_api Docker image using the docker-compose command given in [Local Development](#local-development).
-2. Run `docker run --rm -i notification_api pip freeze > requirements.txt`.
-3. Open requirements.txt, and manually remove any warning messages at the start of the file.
-4. Assuming all unit tests are passing, note any top level dependency updates.  Update requirements-app.txt to make their minimum version equal to the version actually installed according to requirements.txt.
+1. Update ci/Dockerfile.local to use requirements-app.txt, not requirements.txt.
+2. Build the notification_api Docker image using the docker-compose command given in [Local Development](#local-development).
+3. Run `docker run --rm -i notification_api pip freeze > requirements.txt`.
+4. Open requirements.txt, and manually remove any warning messages at the start of the file.
+5. Assuming all unit tests are passing, note any top level dependency updates.  Update requirements-app.txt to make their minimum version equal to the version actually installed according to requirements.txt.
```
Comment (kalbfled) [inline README.md:269]: Step 6: Restore Dockerfile.local to use requirements.txt.

File: requirements-app.txt [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
+botocore >= 1.34.48
+cachelib >= 0.12.0
 cachetools >= 5.3.2
-celery[sqs] >= 5.3.1
+celery[sqs] >= 5.3.6
 click-datetime >= 0.2
 
 # Datadog Tracer and Profiler
-ddtrace >= 1.18.2
+ddtrace == 1.18.2
```
Comment (kalbfled) [inline requirements-app.txt:13]: Why is this being pinned to a specific version in this file?  This means it won't get upgraded during the next routine upgrade.  I think `>=` is correct.

File: requirements-app.txt [hunk 4]
```diff
@@ -19,63 +19,64 @@ eventlet ~= 0.33.3
...
+twilio >= 8.13.0
+Unidecode >= 1.3.8
 validatesns >= 0.1.1
 
 # Flask depends on this package.  It's listed here as a top level dependency--intentionally with no version--because
 # it's directly imported in v3 error handling code to reference the exception BadRequest.
 Werkzeug
+
+# eventlet uses this package, but upgrading past 2.4.2 causes DNS errors.
+dnspython == 2.4.2
```
Comment (kalbfled) [inline requirements-app.txt:82]: Consistent with the rest of the file, please put this in alphabetical order, and date the explanation (i.e. "23 February 2024:").

**Review by kalbfled — CHANGES_REQUESTED** [review]
In the issue comments, please link to the new ticket for upgrading dnspython.

File: requirements-app.txt (hunk 3 above)
Comment (mchlwellman) [reply requirements-app.txt:13]: I was told to handle datadog dependencies in a separate ticket

File: README.md (hunk 2 above)
Comment (mchlwellman) [reply README.md:270]: Ok, I've added that step

File: requirements-app.txt (hunk 4 above)
Comment (mchlwellman) [reply requirements-app.txt]: Ok, I've put dnspython in the correct spot

File: requirements-app.txt (hunk 3 above)
Comment (cris-oddball) [reply requirements-app.txt:13]: Kyle asked him to pin it, it needs to be pinned.

**Review by kalbfled — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Review by MackHalliday — APPROVED** [review]

**Pull request merged**